File: runningmom.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

using namespace std;

#define umap unordered_map

int main() {
    umap<string, vector<string> > adj;
    umap<string, int> inDegree;
    int edges;
    cin >> edges;
    for (int i = 0; i < edges; i++) {
        string src, dest;
        cin >> src >> dest;
        if (adj.count(dest) == 0) {
            vector<string> lst;
            lst.push_back(src);
            adj[dest] = lst;
        } else {
            adj[dest].push_back(src);
        }
        if (inDegree.count(src) == 0) {
            inDegree[src] = 1;
        } else {
            inDegree[src]++;
        } 
        if (inDegree.count(dest) == 0) {
            inDegree[dest] = 0;
        }
    }
    queue<string> q;
    for (auto it : inDegree) {
        if (it.second == 0) {
            q.push(it.first);
        }
    }
    while (!q.empty()) {
        string curr = q.front();
        q.pop();
        inDegree[curr]--;
        for (string s : adj[curr]) {
            inDegree[s]--;
            if (inDegree[s] == 0) {
                q.push(s);
            }
        }
    }
    string query;
    while (cin >> query) {
        if (inDegree[query] > 0) {
            cout << query << " " << "safe" << endl;
        } else {
            cout << query << " " << "trapped" << endl;
        }
    }
}

File: runningmom.py
import collections
import sys

def main():
    flights = int(input())
    inDegree = collections.defaultdict(int)
    adj = collections.defaultdict(list)
    trapped = set()
    for i in range(flights):
        start, end = input().split(" ")
        adj[end].append(start)
        inDegree[start] += 1
        inDegree[end] += 0
    queue = []
    for i in inDegree:
        if inDegree[i] == 0:
            queue.append(i)
    while queue:
        curr = queue.pop()
        for neighbour in adj[curr]:
            inDegree[neighbour] -= 1
            if inDegree[neighbour] == 0:
                queue.append(neighbour)
    for query in sys.stdin:
        query = query.rstrip()
        if inDegree[query] > 0:
            print(query + " safe")
        else:
            print(query + " trapped")

main()